### PR TITLE
Fix bugs that lead to empty display options objects begin created for required display option

### LINKED ISSUE #1228: Internal server error after column type is converted from text to boolean with display options
## To Reproduce
* Create a new text column
* Patch the column from text to boolean with display options.
  ```javascript
  {
    type: 'boolean',
    type_options: {},
    display_options: {
      input: 'dropdown',
      custom_labels: {
        TRUE: 'TRUE'
        FALSE: 'FALSE'
      }
    }
  }
  ```
* Notice that the request succeeds.
* Send a subsequent request to get all columns of the table. Notice the 500 error.
* Try loading the frontend, a debug page with the error message appears.
* **This does not happen always, happens intermittently and I end up having to drop the table directly in the db inorder to get things working again.**

## Error message:
```
[{"code":4999,"message":"\"Got KeyError when attempting to get a value for field `input` on serializer `BooleanDisplayOptionSerializer`.\\nThe serializer field might be named incorrectly and not match any attribute or key on the `dict` instance.\\nOriginal exception text was: 'input'.\"","field":null,"detail":null}]
```

## Error log:
```
Environment:


Request Method: GET
Request URL: http://localhost:8000/mathesar_tables/1/?t=W1tdLG51bGxd

Django Version: 3.1.14
Python Version: 3.9.8
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'rest_framework',
 'django_filters',
 'django_property_filter',
 'mathesar']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']



Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/rest_framework/fields.py", line 457, in get_attribute
    return get_attribute(instance, self.source_attrs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/fields.py", line 95, in get_attribute
    instance = instance[attr]

During handling of the above exception ('input'), another exception occurred:
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/code/mathesar/views.py", line 109, in schema_home
    'common_data': get_common_data(request, database, schema)
  File "/code/mathesar/views.py", line 55, in get_common_data
    'tables': get_table_list(request, schema),
  File "/code/mathesar/views.py", line 35, in get_table_list
    return table_serializer.data
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 745, in data
    ret = super().data
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 246, in data
    self._data = self.to_representation(self.instance)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 663, in to_representation
    return [
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 664, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 515, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 663, in to_representation
    return [
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 664, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/code/mathesar/api/serializers/columns.py", line 67, in to_representation
    return super().to_representation(instance)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 515, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 502, in to_representation
    attribute = field.get_attribute(instance)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/fields.py", line 490, in get_attribute
    raise type(exc)(msg)

Exception Type: KeyError at /mathesar_tables/1/
Exception Value: "Got KeyError when attempting to get a value for field `input` on serializer `BooleanDisplayOptionSerializer`.\nThe serializer field might be named incorrectly and not match any attribute or key on the `dict` instance.\nOriginal exception text was: 'input'."
```

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #1228 

The reason for the error in #1228 is due to an empty `display option` object stored on the database even though certain fields are required, so when the serializer starts looking for the required fields during JSON serialization the error occurs.

I am not able to accurately reproduce the root cause of the empty `display_options` object which leads to the error, but I fixed a  bug that could have led to the creation of empty `display_options`.
- Display_options weren't reevaluated  when partially updating `type` field of a column(`PATCH` /tables/{id}/columns`) which led to the Column model containing the old display options meant for the previous type of a column

 
## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by silentninja** [comment]
@pavish  I am not able to reliably reproduce the issue, Can you confirm if this PR fixes the issue you are facing?

**Comment by pavish** [comment]
@silentninja It was an intermittent issue. I will try to check if it reproduces in this branch, but I cannot confirm for certain.

**Review by kgodey — COMMENTED** [review]
@silentninja This looks good, but I had one question.

File: mathesar/reflection.py [hunk 1]
```diff
@@ -78,6 +78,7 @@ def reflect_columns_from_table(table):
         if not created and column.display_options:
             serializer = DisplayOptionsMappingSerializer(data=column.display_options,
                                                          context={DISPLAY_OPTIONS_SERIALIZER_MAPPING_KEY: str(column.plain_type)})
+            # If the type of column has changed, the display options won't match with existing options
```
Comment (kgodey) [inline mathesar/reflection.py:81]: Can we not invalidate display options during reflection if we detect that they don't match the type?

Comment (silentninja) [reply mathesar/reflection.py:81]: We don't store column type in a Django model, so there is no way to know if the column `type` has changed.

Comment (kgodey) [reply mathesar/reflection.py:81]: We have access to the column type, we can check whether the stored display options make sense for that type. Is there a problem with that approach?

Comment (silentninja) [reply mathesar/reflection.py]: That is what I am doing, I missed out `the valid display options won't match with existing options".

**Review by kgodey — APPROVED** [review]

**Pull request merged**